File: java_solutions_11-20/Euler_14.java
/*
/   Project Euler problem #14 Longest Collatz sequence
/
/   Which starting number, under one million, produces the longest chain?
/
*/


public class Euler_14
{
    public static void main(String[] args)
    {
        int longest_chain = 0;
        int high = 0;

        for (int current_number = 0; current_number < 1000000; current_number++) {

            int chain = 1;
            long chain_number = current_number;

            while (chain_number > 1) {

                if (chain_number % 2 == 0) {
                    chain_number = chain_number / 2;
                } else {
                    chain_number = (chain_number * 3) + 1;
                }

                chain = chain + 1;
            }

            if (chain > longest_chain) {
                longest_chain = chain;
                high = current_number;
            }
        }
            
        System.out.printf("Answer: %d\n", high);   
    }
}

File: Python/py_solutions_11-20/Euler_14.py
# -*- coding: utf-8 -*-

from __future__ import print_function

'''
The following iterative sequence is defined for the set of positive integers:

n → n/2 (n is even)
n → 3n + 1 (n is odd)

Using the rule above and starting with 13, we generate the following sequence:
13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1

It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. 
Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.
'''

import timeit


try:
    xrange = range
except NameError:
    pass

start = timeit.default_timer()

def collatz(n):
    count = 0
    while n != 1:
        if n & 1:
            n *= 3 
            n += 1
            count += 1
        if not n & 1:
            n >>= 1 
            count += 1
    return count 

def euler_14():
    canidates = xrange(3, 1000000, 2)
    number_sequences = {collatz(i):i for i in canidates}
    return number_sequences[max(number_sequences)]

print("Answer: {}".format(euler_14()))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format((stop - start)))
